test/lint/lint-circular-dependencies.sh

```diff
@@ -0,0 +1,145 @@
+    "chain -> legacy/stakemodifier -> validation -> evo/specialtx -> chain"
+    "chain -> legacy/stakemodifier -> validation -> validationinterface -> chain"
+    "chain -> legacy/stakemodifier -> stakeinput -> txdb -> chain"
+    "chain -> legacy/stakemodifier -> validation -> undo -> chain"
+    "chain -> legacy/stakemodifier -> validation -> zpivchain -> chain"
+    "chainparams -> checkpoints -> validation -> consensus/tx_verify -> chainparams"
+    "chainparams -> checkpoints -> validation -> evo/specialtx -> chainparams"
+    "chainparams -> checkpoints -> validation -> pow -> chainparams"
+    "chainparams -> checkpoints -> validation -> zpiv/zpivmodule -> chainparams"
+    "coins -> policy/fees -> policy/policy -> validation -> coins"
+    "consensus/tx_verify -> consensus/zerocoin_verify -> spork -> validation -> consensus/tx_verify"
+    "addrman -> timedata -> chainparams -> checkpoints -> validation -> addrman"
+    "blocksignature -> primitives/block -> script/sign -> policy/policy -> validation -> blocksignature"
+    "coins -> policy/fees -> policy/policy -> validation -> txdb -> coins"
+    "consensus/tx_verify -> evo/specialtx -> primitives/block -> script/sign -> policy/policy -> consensus/tx_verify"
+    "consensus/merkle -> primitives/block -> script/sign -> policy/policy -> validation -> consensus/merkle"
+    "evo/specialtx -> primitives/block -> script/sign -> policy/policy -> validation -> evo/specialtx"
+    "policy/policy -> validation -> pow -> primitives/block -> script/sign -> policy/policy"
+    "chain -> legacy/stakemodifier -> stakeinput -> wallet/wallet -> spork -> net_processing -> chain"
+)
+
+EXIT_CODE=0
+
+CIRCULAR_DEPENDENCIES=()
+
+IFS=$'\n'
+for CIRC in $(cd src && ../contrib/devtools/circular-dependencies.py {*,*/*,*/*/*}.{h,cpp} | sed -e 's/^Circular dependency: //'); do
+    CIRCULAR_DEPENDENCIES+=( "$CIRC" )
+    IS_EXPECTED_CIRC=0
+    for EXPECTED_CIRC in "${EXPECTED_CIRCULAR_DEPENDENCIES[@]}"; do
+        if [[ "${CIRC}" == "${EXPECTED_CIRC}" ]]; then
+            IS_EXPECTED_CIRC=1
+            break
+        fi
+    done
+    if [[ ${IS_EXPECTED_CIRC} == 0 ]]; then
+        echo "A new circular dependency in the form of \"${CIRC}\" appears to have been introduced."
+        echo
+        EXIT_CODE=1
+    fi
+done
+
+for EXPECTED_CIRC in "${EXPECTED_CIRCULAR_DEPENDENCIES[@]}"; do
+    IS_PRESENT_EXPECTED_CIRC=0
+    for CIRC in "${CIRCULAR_DEPENDENCIES[@]}"; do
+        if [[ "${CIRC}" == "${EXPECTED_CIRC}" ]]; then
+            IS_PRESENT_EXPECTED_CIRC=1
+            break
+        fi
+    done
+    if [[ ${IS_PRESENT_EXPECTED_CIRC} == 0 ]]; then
+        echo "Good job! The circular dependency \"${EXPECTED_CIRC}\" is no longer present."
+        echo "Please remove it from EXPECTED_CIRCULAR_DEPENDENCIES in $0"
+        echo "to make sure this circular dependency is not accidentally reintroduced."
+        echo
+        EXIT_CODE=1
+    fi
+done
+
+exit ${EXIT_CODE}
```

Comment:
tiny nit: jump line at the end of the file.

Reply:
:+1: will add the jump line in either #2643, #2644 or #2646 (which are all editing this file, and will all need to be rebased when this PR is merged).